benchmark/solitary_wave/solitary_wave.cc

```diff
@@ -0,0 +1,1021 @@
+#include <aspect/melt.h>
+#include <aspect/compositional_initial_conditions/interface.h>
+#include <aspect/simulator_access.h>
+#include <aspect/global.h>
+
+#include <deal.II/dofs/dof_tools.h>
+#include <deal.II/dofs/dof_handler.h>
+#include <deal.II/numerics/data_out.h>
+#include <deal.II/base/quadrature_lib.h>
+#include <deal.II/base/function_lib.h>
+#include <deal.II/numerics/error_estimator.h>
+#include <deal.II/numerics/vector_tools.h>
+
+
+/**
+ * Computes MPI_MAX of @p local_max like Allreduce, but also transmits the @p local_data from the rank with the largest @local_max to every rank (returned in @p global_data).
+ */
+template <class T1, class T2>
+void MPI_max_and_data(const T1 &local_max,
+                      const T2 &local_data,
+                      T1 &global_max,
+                      T2 &global_data)
+{
+  AssertThrow(false, dealii::ExcNotImplemented());
+}
+
+void myop_func(void *invec,
+               void *inoutvec,
+               int *len,
+               MPI_Datatype *datatype)
+{
+  AssertThrow(*len > 1, dealii::ExcNotImplemented());
+  AssertThrow(*datatype == MPI_DOUBLE, dealii::ExcNotImplemented());
+
+  double *indata = static_cast<double *>(invec);
+  double *inoutdata = static_cast<double *>(inoutvec);
+
+  if (indata[0]>inoutdata[0])
+    {
+      for (int i=0; i<*len; ++i)
+        inoutdata[i] = indata[i];
+    }
+}
+
+
+template <>
+void MPI_max_and_data(const double &local_max,
```

Comment:
if all you need is the specialization for `T=double`, why declare the general template above at all?